### ISSUE TITLE
Prediction and ground truth allignment in loss functions

### ISSUE THREAD

**Issue by dutchJSCOOP** [matterport/Mask_RCNN#1218]
In the mrcnn_mask_loss_graph function, the y_pred and y_true are used to calculate the binary cross-entropy loss. I am curious how these are obtained. The pred_masks input has for each proposal a 3d matrix (h,w, n_classes) resulting from a sigmoid layer in build_fpn_mask_graph (shouldn't you use softmax for more than 2 classes?). 

y_pred is then obtained by gathering the masks with the target_class_ids. So the assumption is made that the class  (/relevant sigmoid output slice) from the i'th proposal is in fact the ground truth class of the i'th target_mask. Does this mean the pred_masks and the target_masks are aligned somehow? How can you, for instance, assume that the first target_mask corresponds to the first proposal? I.o.w. that if the first target_mask has class one, that the first proposal then also must be class one. I need this information for developing [this AP metric](http://cocodataset.org/#detection-eval) (which I intend to share if it works), where I need to match proposals and targets by means of max IOU. 

I probably misunderstand what is happening, but any help would be greatly appreciated.
